prow/cmd/build/controller.go

```diff
@@ -576,6 +579,62 @@ func injectSource(b *buildv1alpha1.Build, pj prowjobv1.ProwJob) error {
 		b.Spec.Template.Arguments = append(b.Spec.Template.Arguments, wd)
 	}
 
+	return true, nil
+}
+
+func tools() (v1.Volume, v1.VolumeMount) {
+	const toolsName = "entrypoint-tools"
+	toolsVolume := v1.Volume{
+		Name: toolsName,
+		VolumeSource: v1.VolumeSource{
+			EmptyDir: &v1.EmptyDirVolumeSource{},
+		},
+	}
+
+	toolsMount := v1.VolumeMount{
+		Name:      toolsName,
+		MountPath: "/entrypoint-tools",
+	}
+	return toolsVolume, toolsMount
+}
+
+func decorateBuild(b *buildv1alpha1.Build, encodedJobSpec string, dc prowjobv1.DecorationConfig, injectedSource bool) error {
+	toolsVolume, toolsMount := tools()
+	gcsVol, gcsMount, gcsOptions := decorate.GCSOptions(dc)
+
+	var entries []wrapper.Options
+	for i := range b.Spec.Steps {
+		if b.Spec.Steps[i].Name == "" {
+			b.Spec.Steps[i].Name = fmt.Sprintf("step %d", i)
+		}
+		alwaysPass := true // last step is sidecar, which is what should exit non-zero
+		var previousMarker string
+		if i > 0 {
+			previousMarker = entries[i-1].MarkerFile
+		}
+		opt, err := decorate.InjectEntrypoint(&b.Spec.Steps[i], dc.Timeout, dc.GracePeriod, b.Spec.Steps[i].Name, previousMarker, alwaysPass, logMount, toolsMount)
```

Comment:
Looks like you just added a comment? Are we just going to give the full timeout to each step initially?

Reply:
See L641 in `decorateBuild()` where I now set the build's timeout.

Reply:
Ok, so we set each step to have the full timeout and also set the build as a whole to have the timeout? The build timeout will be stricter since the entrypoint timeouts for each step start counting when the step starts. If the build as a whole times out we will fail to upload results, right?
Using a deadline instead of a duration like you mentioned in the comment would make it easy to define a timeout across steps.

Reply:
Right, not perfect. IMO something simple good enough for now that we can improve later.